Make container runtime's cgroup configurable.

vishh opened the pull request:
Use the real cgroups for metrics generation.

With this PR, on systemd systems, users can disable most of the cgroups related configuration that kubelet does, and still have the metrics APIs work for them (mostly).

cc @pwittrock @dchen1107 

I couldn't resist myself from cleaning up the code along the way.

vishh committed:
Make container runtime's cgroup configurable.
Use the real cgroups for metrics generation.

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
cc @derekwaynecarr

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -216,4 +215,5 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.OutOfDiskTransitionFrequency.Duration, "outofdisk-transition-frequency", s.OutOfDiskTransitionFrequency.Duration, "Duration for which the kubelet has to wait before transitioning out of out-of-disk node condition status. Default: 5m0s")
 	fs.StringVar(&s.NodeIP, "node-ip", s.NodeIP, "IP address of the node. If set, kubelet will use this IP address for the node")
 	fs.BoolVar(&s.EnableCustomMetrics, "enable-custom-metrics", s.EnableCustomMetrics, "Support for gathering custom metrics.")
+	fs.StringVar(&s.RuntimeContainer, "runtime-container", s.RuntimeContainer, "Absolute name of the cgroups to create (if required) and run the runtime in (Default: /docker-daemon).")
```
I would prefer we don't expose this to the user.  How do you see this mapping to when we enforce --system-reserved and --kube-reserved?

vishh replied:
On systemd nodes, this flag can be set to `""` which will prevent kubelet from moving docker daemon into `/docker-daemon` cgroup. Similarly, if users want a custom configuration, they can have one, without breaking rest of kubelet. WDYT?

derekwaynecarr replied:
I prefer the model in my PR.   I don't like a default being wrong for a huge portion of users running on systemd environments so I would prefer we discover resource container or ensure its a child of system container or Kubelet container.

vishh replied:
This PR ensures that we will discover the docker daemon's container, if this flag is set to `""`. So thats taken care of.
The only question that still left is that of defaults. There are at-least two sets of users - systemd, non-systemd (maybe more). Each party wants to setup the node differently. We can either recognize each type of configuration as a first-class entity in kubelet, or let users express their node configuration explicitly.
If we go with the former, we will have a built-in policy for systemd, one for GCP, and possible more.
With this model, we will publish a cookbook and avoid needing inbuilt models. WDYT?

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -114,6 +112,7 @@ func NewKubeletServer() *KubeletServer {
 			RktPath:                        "",
 			RktStage1Image:                 "",
 			RootDirectory:                  defaultRootDir,
+			RuntimeContainer:               "/docker-daemon",
```
The assumption is that this fixes systemd if and only if you set this flag to "/system.slice/docker.service"?

  It looks like the default value is always wrong for systemd.  Maybe I need to look more as I review PR

vishh replied:
All the container defaults are wrong for systemd. They have to be set to `""` as I mentioned above.

derekwaynecarr replied:
If you can review my PR on Monday, maybe we can find a suitable middle ground but I don't like exposing yet another container name for users to manage.

vishh replied:
The issue right now is that there is no easy way for people to restrict the kubelet from not moving docker daemon into `/docker-daemon` cgroups. This PR fixes that in addition to fixing the metrics integration.
Magical names are much harder to manager. If we are going with default cgroup hierarchies, I'd prefer getting rid of all the other flags that let users specify cgroups and let kubelet decide defaults - be it systemd or non-systemd nodes.

derekwaynecarr commented:
I prefer a model where if systemd is detected, we discover the RuntimeContainer and ignore the user specified value altogether.   I also prefer we not expose RuntimeContainer since I think it complicates enforcement of kube-reserved or system-reserved since it may not map to either --system-container-name or --resource-container-name.  It's late on Friday, so apologies if I am missing something.

vishh commented:
Auto-detection is possible, but given that we already have setups were
systemd is not running as init (yes it does exist), we will have to work
around that as well. Hence the reason for explicit configuration.

Here is my understanding:
`kube-reserved` should map to `--resource-container`, which is set to
`/kubelet` by default. On systemd nodes, its value should be a different
slice.
`system-reserved` should map to `--system-container` which is set to `""`
by default. On systemd nodes, this should map to `system.slice` I think.

Where does `--runtime-container` fit in this model? Is it part of
`system-reserved` or `kube-reserved`? Isn't that deployment specific?


On Fri, Feb 5, 2016 at 3:32 PM, Derek Carr <notifications@github.com> wrote:

> I prefer a model where if systemd is detected, we discover the
> RuntimeContainer and ignore the user specified value altogether. I also
> prefer we not expose RuntimeContainer since I think it complicates
> enforcement of kube-reserved or system-reserved since it may not map to
> either --system-container-name or --resource-container-name. It's late on
> Friday, so apologies if I am missing something.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180620943>
> .
>

vishh commented:
Apologies for not having cc'ed you in this PR earlier @derekwaynecarr.

dchen1107 commented:
Vish, I am confused. If we have to set all flags to "" for systemd node here, where the processes are end up to at cgorup hierarchy here? I like @derekwaynecarr's  pr (#20748) better since it is aligned with the systemd nodespec which we agreed upon. Of course on another side, I do agree with you that if we want kubelet auto discover the systemd node, and set the default, we should do the same to the non-systemd node. But at least I think his approach is moving further. 

I need to read both prs more in detail.

vishh commented:
If `--runtime-container=""` and `--resource-container=""`, kubelet will
auto-discover the cgroups of docker daemon and kubelet and use that cgroup
for node metrics purposes.


On Fri, Feb 5, 2016 at 4:20 PM, Dawn Chen <notifications@github.com> wrote:

> Vish, I am confused. If we have to set all flags to "" for systemd node
> here, where the processes are end up to at cgorup hierarchy here? I like
> @derekwaynecarr <https://github.com/derekwaynecarr>'s pr (#20748
> <https://github.com/kubernetes/kubernetes/pull/20748>) better since it is
> aligned with the systemd nodespec which we agreed upon. Of course on
> another side, I do agree with you that if we want kubelet auto discover the
> systemd node, and set the default, we should do the same to the non-systemd
> node. But at least I think his approach is moving further.
>
> I need to read both prs more in detail.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180632501>
> .
>

derekwaynecarr commented:
We discussed that at F2F and Dawn wanted it tracked under kube-reserved,
and I wanted it tracked under system-reserved. We agreed to not mandate
either, but we did not want to have a third container to manage specific to
the runtime.

On Friday, February 5, 2016, Vish Kannan <notifications@github.com> wrote:

> Auto-detection is possible, but given that we already have setups were
> systemd is not running as init (yes it does exist), we will have to work
> around that as well. Hence the reason for explicit configuration.
>
> Here is my understanding:
> `kube-reserved` should map to `--resource-container`, which is set to
> `/kubelet` by default. On systemd nodes, its value should be a different
> slice.
> `system-reserved` should map to `--system-container` which is set to `""`
> by default. On systemd nodes, this should map to `system.slice` I think.
>
> Where does `--runtime-container` fit in this model? Is it part of
> `system-reserved` or `kube-reserved`? Isn't that deployment specific?
>
>
> On Fri, Feb 5, 2016 at 3:32 PM, Derek Carr <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > I prefer a model where if systemd is detected, we discover the
> > RuntimeContainer and ignore the user specified value altogether. I also
> > prefer we not expose RuntimeContainer since I think it complicates
> > enforcement of kube-reserved or system-reserved since it may not map to
> > either --system-container-name or --resource-container-name. It's late on
> > Friday, so apologies if I am missing something.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180620943
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180625422>
> .
>

vishh commented:
Yeah. Thats what I recall too from the F2F, but is there a reason why we
want to not have explicit config?
I'm with you on not adding additional flags as much as possible, but in
this case, I feel that magic defaults are not good. I'm all ears for
alternative approaches.

On Fri, Feb 5, 2016 at 4:27 PM, Derek Carr <notifications@github.com> wrote:

> We discussed that at F2F and Dawn wanted it tracked under kube-reserved,
> and I wanted it tracked under system-reserved. We agreed to not mandate
> either, but we did not want to have a third container to manage specific to
> the runtime.
>
> On Friday, February 5, 2016, Vish Kannan <notifications@github.com> wrote:
>
> > Auto-detection is possible, but given that we already have setups were
> > systemd is not running as init (yes it does exist), we will have to work
> > around that as well. Hence the reason for explicit configuration.
> >
> > Here is my understanding:
> > `kube-reserved` should map to `--resource-container`, which is set to
> > `/kubelet` by default. On systemd nodes, its value should be a different
> > slice.
> > `system-reserved` should map to `--system-container` which is set to `""`
> > by default. On systemd nodes, this should map to `system.slice` I think.
> >
> > Where does `--runtime-container` fit in this model? Is it part of
> > `system-reserved` or `kube-reserved`? Isn't that deployment specific?
> >
> >
> > On Fri, Feb 5, 2016 at 3:32 PM, Derek Carr <notifications@github.com
> > <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> >
> > > I prefer a model where if systemd is detected, we discover the
> > > RuntimeContainer and ignore the user specified value altogether. I also
> > > prefer we not expose RuntimeContainer since I think it complicates
> > > enforcement of kube-reserved or system-reserved since it may not map to
> > > either --system-container-name or --resource-container-name. It's late
> on
> > > Friday, so apologies if I am missing something.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180620943
> > >
> > > .
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180625422
> >
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-180633565>
> .
>

vishh commented:
To summarize, this PR helps make kubelet not update cgroups, if required. For example, this PR will help make local kube cluster not move docker daemon into a different cgroup.
In addition to that, this PR makes the metrics API `/stats/summary` report metrics for system containers even if the containers were not created by kubelet.

@derekwaynecarr @dchen1107: Can we postpone the discussion of how to handle systemd vs other setups and node allocable to #20748?

I'd like to include this PR in v1.2 so that the metrics API works on all setups.

derekwaynecarr commented:
How do you feel about changing the default value of the runtime container
flag?  The current behavior was wrong on a huge number of setups and I
would prefer we do not carry the wrong behavior as the flag default.  The
default behavior in my opinion should be do nothing with docker unless
told.  I also continue to not understand why we would have the default not
align with system or Kubelet containers.

On Monday, February 8, 2016, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e test build/test *passed* for commit 0f69860
> <https://github.com/kubernetes/kubernetes/commit/0f69860c12fcbd66cae4ba5bafc118f93a0ac327>
> .
>
>    - Build Log
>    <http://pr-test.k8s.io/20687/kubernetes-pull-build-test-e2e-gce/27547/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/20687/kubernetes-pull-build-test-e2e-gce/27547/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//27547>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-181666112>
> .
>

derekwaynecarr commented:
@vishh - if you want to meet today in a hangout or something to discuss way forward, let me know.  I generally view the current behavior as an undocumented bug, and my main concern with this PR is that it makes that bug the default behavior.  I also think the --runtime-container default setting should be a child of system or kube container.

vishh commented:
Let's chat over VC. I'd be down to changing the default. I agree with you that the default doesn't make sense everywhere. 
Other than that any other issues with this PR?

dchen1107 commented:
Looks like we are converging on this:

* Vish is going to change the default value of the runtime container through the newly introduced flag. 
* @derekwaynecarr you are right on that default setting of docker should be child of system or kube container in our NodeAllocatable design. The only difference is that in your environment, you prefer to put it into system; and in our current environment, we would rather have them in kube. 
* In 1.2 release, we are measuring both kubelet and docker's resource usages. If we can put the reservation for kube container to make NodeAllocatable more meaningful and prevent bad overcommit, that would be great.
* We don't expect to enforce those limit on the node in 1.2 release timeframe.

vishh committed:
Updating the default values for the cgroups hierarchies that kubelet
creates for managing the nodes.

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
@dchen1107 @derekwaynecarr: Updated the default values. PTAL

vishh commented on cluster/saltbase/salt/kubelet/default:
```diff
@@ -102,15 +102,19 @@
     {% set cgroup_root = "--cgroup-root=docker" -%}
   {% else %}
     {% set cgroup_root = "--cgroup-root=/" -%}
+    {% set system_container = "--system-container=/system" -%}    
```
cc @roberthbailey: We will have to mirror these flags in GKE.

dchen1107 replied:
GKE will pick up those flags through salt when the node coming up today. Nothing is different here.

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -218,4 +217,5 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.OutOfDiskTransitionFrequency.Duration, "outofdisk-transition-frequency", s.OutOfDiskTransitionFrequency.Duration, "Duration for which the kubelet has to wait before transitioning out of out-of-disk node condition status. Default: 5m0s")
 	fs.StringVar(&s.NodeIP, "node-ip", s.NodeIP, "IP address of the node. If set, kubelet will use this IP address for the node")
 	fs.BoolVar(&s.EnableCustomMetrics, "enable-custom-metrics", s.EnableCustomMetrics, "Support for gathering custom metrics.")
+	fs.StringVar(&s.RuntimeContainer, "runtime-container", s.RuntimeContainer, "Optional absolute name of the cgroups to create and run the runtime in.")
```
cgroups is plural here where it should be singular.

vishh replied:
Updating.

derekwaynecarr commented on cmd/kubelet/app/server.go:
```diff
@@ -493,7 +502,7 @@ func SimpleKubelet(client *clientset.Clientset,
 		DiskSpacePolicy:           diskSpacePolicy,
 		DockerClient:              dockerClient,
-		DockerDaemonContainer:     "/docker-daemon",
+		RuntimeContainer:          "/docker-daemon",
```
Can we set this value as empty string?  I don't want integration tests to try and move my docker daemon...

vishh replied:
Ahh missed this. Thanks for pointing out.

derekwaynecarr commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -25,15 +25,19 @@ type ContainerManager interface {
 	SystemContainersLimit() api.ResourceList
+
+	// Returns a NodeConfig that is being used by the container manager.
+	GetNodeConfig() NodeConfig
```
+1 to this change

derekwaynecarr commented:
This is much improved, I had a few minor nits if you can update, then it would be lgtm.

For kube 1.3, we should separate the convention of what cgroup we place pids in versus what cgroup we enforce the node constraints against.  For example, on systemd, I want to say that the system container is "/system.slice" and that is where we should enforce --system-reserved, but I do not want to have the kubelet move any pids.  In general, I never want to move pids on any systemd system.  That may require us to have additional flags, like `--system-reserved-cgroup`.

vishh commented:
Addressed comments. 
We can add enforcement of node allocable values in a separate PR.
Moving pids has always been a short-term solution. I'm considering making that behavior opt-in, possibly in a future PR, after v1.2 branch is cut.

derekwaynecarr commented:
@vishh @dchen1107 LGTM

dchen1107 commented on cluster/gce/trusty/master.yaml:
```diff
@@ -154,6 +154,8 @@ script
 		--configure-cbr0=${ALLOCATE_NODE_CIDRS} \
 		--cgroup-root=/ \
 		--system-container=/system \
+                --runtime-container=/docker-daemon \
```
@andyzheng0831  FYI

andyzheng0831 replied:
Please use tab instead of spaces and fix the indentation

andyzheng0831 commented on cluster/gce/trusty/node.yaml:
```diff
@@ -139,6 +139,8 @@ script
 		--cgroup-root=/ \
 		--system-container=/system \
+                --runtime-container=/docker-daemon \
+                --resource-container=/kubelet \
```
Use tabs instead of spaces

vishh replied:
Why aren't we using spaces instead of tabs? Spaces are portable across editors. I can update tabs to be spaces in the entire file if that's ok with you @andyzheng0831.

dchen1107 replied:
If I understood correctly, the tab is required by whatever version of cloud_init version they are using. But will leave @andyzheng0831 for last answer.

andyzheng0831 replied:
See here: http://upstart.ubuntu.com/cookbook/#code-style. Using tabs is the code style of upstart jobs.

dchen1107 commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -25,15 +25,19 @@ type ContainerManager interface {
-	KubeletContainerName      string
+	RuntimeContainerName string
+	SystemContainerName  string
+	KubeletContainerName string
```
s/KubeletContainerName/ResourceContainerName to keep the name consistent with the flag? Or change the flag name to kubelet-container?

vishh replied:
Ideally, the flag name has to change. Don't we have a flag deprecation
policy? Can I just rename a flag? With this change, the internal code is
more readable though.

On Tue, Feb 9, 2016 at 3:30 PM, Dawn Chen <notifications@github.com> wrote:

> In pkg/kubelet/cm/container_manager.go
> <https://github.com/kubernetes/kubernetes/pull/20687#discussion_r52392670>
> :
>
> >  }
> >
> >  type NodeConfig struct {
> > -	DockerDaemonContainerName string
> > -	SystemContainerName       string
> > -	KubeletContainerName      string
> > +	RuntimeContainerName string
> > +	SystemContainerName  string
> > +	KubeletContainerName string
>
> s/KubeletContainerName/ResourceContainerName to keep the name consistent
> with the flag? Or change the flag name to kubelet-container?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687/files#r52392670>.
>

derekwaynecarr replied:
I find KubeletContainerName much easier to understand than the old label.
Can we mark the resource container name flag deprecated and just have
Kubelet container name?  Until flag is removed we could keep both values in
synch?

On Tuesday, February 9, 2016, Vish Kannan <notifications@github.com> wrote:

> In pkg/kubelet/cm/container_manager.go
> <https://github.com/kubernetes/kubernetes/pull/20687#discussion_r52393199>
> :
>
> >  }
> >
> >  type NodeConfig struct {
> > -	DockerDaemonContainerName string
> > -	SystemContainerName       string
> > -	KubeletContainerName      string
> > +	RuntimeContainerName string
> > +	SystemContainerName  string
> > +	KubeletContainerName string
>
> Ideally, the flag name has to change. Don't we have a flag deprecation
> policy? Can I just rename a flag? With this change, the internal code is
> more readable though.
> … <#-1321304034_>
> On Tue, Feb 9, 2016 at 3:30 PM, Dawn Chen <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote: In
> pkg/kubelet/cm/container_manager.go <#20687 (comment)
> <https://github.com/kubernetes/kubernetes/pull/20687#discussion_r52392670>>
> : > } > > type NodeConfig struct { > - DockerDaemonContainerName string > -
> SystemContainerName string > - KubeletContainerName string > +
> RuntimeContainerName string > + SystemContainerName string > +
> KubeletContainerName string s/KubeletContainerName/ResourceContainerName to
> keep the name consistent with the flag? Or change the flag name to
> kubelet-container? — Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/20687/files#r52392670>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687/files#r52393199>.
>

dchen1107 replied:
+1 on KubeletContainerName

vishh replied:
I'm going to name it `--kubelet-cgroups` to be aligned with the allocable work in the future.
In retrospect, I want to rename `--runtime-container` to be `--runtime-cgroups`, and `--system-container` to `--system-cgroups`. The keyword `container` will confuse users.

dchen1107 replied:
Changing to *-groups sounds good to me.

vishh replied:
I went ahead and renamed flags and variables.

dchen1107 commented:
LGTM except small nits.

Also you have to rerun hack/verify-generated-docs.sh and fix those mismatches.

andyzheng0831 commented:
Just notice that you change all tabs in cluster/gce/trusty/master.yaml and node.yaml to spaces. It is upstart's code style of using tabs (http://upstart.ubuntu.com/cookbook/#code-style). Please correct your changes to the tabs in these two files.

vishh committed:
Replace `--resource-container` and `--system-container` with
`--kubelet-cgroups` and `--system-cgroups` respectively.
Updated `--runtime-container` to `--runtime-cgroups`.
Cleaned up most of the kubelet code that consumes these flags to match
the flag name changes.

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh committed:
Update flag values for cgroups managed by kubelet.

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
@andyzheng0831: Reverted to tabs
@dchen1107: PTAL

dchen1107 commented on pkg/kubelet/server/stats/summary_test.go:
```diff
@@ -48,9 +48,9 @@ func TestBuildSummary(t *testing.T) {
-		DockerDaemonContainerName: "/docker-daemon",
-		SystemContainerName:       "/system",
-		KubeletContainerName:      "/kubelet",
+		RuntimeContainerName: "/docker-daemon",
```
Shouldn't your rename this to RuntimeCgroupsName?

vishh replied:
Yeah. Fixing it now.

dchen1107 commented on pkg/kubelet/server/stats/summary_test.go:
```diff
@@ -48,9 +48,9 @@ func TestBuildSummary(t *testing.T) {
-		SystemContainerName:       "/system",
-		KubeletContainerName:      "/kubelet",
+		RuntimeContainerName: "/docker-daemon",
+		SystemContainerName:  "/system",
```
SystemCgroupsName here?

vishh replied:
Ditto.

dchen1107 commented:
Compiling issue: 

# k8s.io/kubernetes/contrib/mesos/pkg/executor/service
/go/src/k8s.io/kubernetes/contrib/mesos/pkg/executor/service/service.go:193: kcfg.ResourceCgroups undefined (type *app.KubeletConfig has no field or method ResourceCgroups)

andyzheng0831 commented:
LGTM for the change to cluster/gce/trusty files

dchen1107 commented:
@vishh I still see one unittest failed here:

FAIL	k8s.io/kubernetes/pkg/apis/componentconfig/install	0.030s

vishh commented:
Yeah. Fixing generated code.

On Wed, Feb 10, 2016 at 2:40 PM, Dawn Chen <notifications@github.com> wrote:

> @vishh <https://github.com/vishh> I still see one unittest failed here:
>
> FAIL k8s.io/kubernetes/pkg/apis/componentconfig/install 0.030s
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20687#issuecomment-182614294>
> .
>

dchen1107 commented:
Unittest are still failed:

FAIL	k8s.io/kubernetes/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned	0.181s
FAIL	k8s.io/kubernetes/pkg/storage/etcd	7.963s

dchen1107 commented:
LGTM

xiaoanyunfei commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -192,70 +192,113 @@ func (cm *containerManagerImpl) setupNode() error {
-
-		err := ensureSystemContainer(rootContainer, manager)
+		cont.ensureStateFunc = func(_ *fs.Manager) error {
+			return manager.Apply(os.Getpid())
```
Why create a new  `fs.Manager` instead of the default `cont.manager`?
